nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/QueryAzureDataExplorer.java

```diff
@@ -0,0 +1,176 @@
+    private final List<PropertyDescriptor> descriptors = List.of(ADX_SOURCE_SERVICE,DB_NAME,ADX_QUERY);
+    private AdxSourceConnectionService service;
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return this.relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        service = context.getProperty(ADX_SOURCE_SERVICE).asControllerService(AdxSourceConnectionService.class);
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile outgoingFlowFile;
+        String databaseName = context.getProperty(DB_NAME).getValue();
+        String adxQuery;
+        KustoQueryResponse kustoQueryResponse;
+
+        //checks if this processor has any preceding connection, if yes retrieve
+        if (context.hasIncomingConnection()) {
+            FlowFile incomingFlowFile = session.get();
+            //incoming connection exists but the incoming flowfile is null
+            if (incomingFlowFile == null && context.hasNonLoopConnection()) {
+                return;
+            }
+            //incoming connection exists and retrieve adxQuery from context
+            if (incomingFlowFile != null && incomingFlowFile.getSize() == 0) {
+                if (context.getProperty(ADX_QUERY).isSet()) {
+                    adxQuery = context.getProperty(ADX_QUERY).evaluateAttributeExpressions(incomingFlowFile).getValue();
+                } else {
+                    String message = "FlowFile query is empty and no scheduled query is set";
+                    getLogger().error(message);
+                    incomingFlowFile = session.putAttribute(incomingFlowFile, ADX_QUERY_ERROR_MESSAGE, message);
+                    session.transfer(incomingFlowFile, FAILED);
+                    return;
+                }
+            } else {
+                try {
+                    adxQuery = getQuery(session, incomingFlowFile);
+                } catch(IOException ioe) {
+                    throw new ProcessException("Failed to read Query from FlowFile",ioe);
+                }
+            }
+            outgoingFlowFile = incomingFlowFile;
+        } else {
+            outgoingFlowFile = session.create();
+            adxQuery = context.getProperty(ADX_QUERY).evaluateAttributeExpressions(outgoingFlowFile).getValue();
+        }
+
+        try {
+            //execute Query
+            kustoQueryResponse = executeQuery(databaseName,adxQuery);
+            if(!kustoQueryResponse.isError()){
+                try(ByteArrayInputStream bais = new ByteArrayInputStream(objectMapper.writeValueAsBytes(kustoQueryResponse.getTableData()))){
+                    session.importFrom(bais, outgoingFlowFile);
+                }
```

Comment:
This approach is very memory-intensive and needs to be changed. Serializing the entire result set into a byte array can easily lead to memory exhaustion for large numbers of results. The Kusto ClientFactory also includes a `StreamingClient` that returns an `InputStream`. Based on that option, both the Controller Service interface and this Processor implementation should be changed. The Service should return a streaming query result, which can be transferred directly to the FlowFile OutputStream using one of the methods on ProcessSession.

Reply:
makes sense @exceptionfactory, fixed as advised